csvcubed/csvcubed/models/cube/qb/components/codelist.py

```diff
@@ -159,8 +157,28 @@ def get_permitted_rdf_fragment_hints(self) -> Set[RdfSerialisationHint]:
     def get_default_node_serialisation_hint(self) -> RdfSerialisationHint:
         return RdfSerialisationHint.ConceptScheme
 
-    def validate_data(self, data: PandasDataTypes) -> List[ValidationError]:
-        return []  # TODO: implement this.
+    def validate_data(self, data: PandasDataTypes) -> list[Optional[ValidationError]]:
+        """
+        Validate the data held in the codelists, assuming case insensitivity
+        """
+
+        errors = list()
+
+        if data is not None:
+            unique_keys = dict(list)
+
+            # Create a dict of all uri-safe values with the originals as a list
+            for k, v in {value: uri_safe(value) for value in data.unique()}.items():
+                unique_keys[k].setdefault(v, []).append(k)
+
+            # Check for collisions, raise a Validation error
+            for k, v in unique_keys:
+                if len(v) > 1:
+                    errors += ValidationError(
+                        message=f'Labels "{v}" collide as single uri-safe value "{k}"'
```

Comment:
This won't be the most user-friendly error message. I'm imagining `Labels "["One", "Two"]" collide.... `. It's okay for now, but we'll need to improve later.